linker/Linker.Steps/CodeRewriterStep.cs

```diff
@@ -0,0 +1,127 @@
+		{
+			if (Annotations.GetAction (assembly) != AssemblyAction.Link)
+				return;
+
+			this.assembly = assembly;
+
+			foreach (var type in assembly.MainModule.Types)
+				ProcessType (type);
+		}
+
+		void ProcessType (TypeDefinition type)
+		{
+			foreach (var method in type.Methods) {
+				if (method.HasBody)
+					ProcessMethod (method);
+			}
+
+			foreach (var nested in type.NestedTypes)
+				ProcessType (nested);
+		}
+
+		void ProcessMethod (MethodDefinition method)
+		{
+			switch (Annotations.GetAction (method)) {
+			case MethodAction.ConvertToStub:
+				RewriteBodyToStub (method);
+				break;
+			case MethodAction.ConvertToThrow:
+				RewriteBodyToLinkedAway (method);
+				break;
+			}
+		}
+
+		void RewriteBodyToLinkedAway (MethodDefinition method)
+		{
+			method.ImplAttributes &= ~(MethodImplAttributes.AggressiveInlining | MethodImplAttributes.Synchronized);
+			method.ImplAttributes |= MethodImplAttributes.NoInlining;
+
+			method.Body = CreateThrowLinkedAwayBody (method);
+			ClearDebugInformation (method);
+		}
+
+		void RewriteBodyToStub (MethodDefinition method)
+		{
+			method.Body = CreateStubBody (method);
+
+			ClearDebugInformation (method);
+		}
+
+		MethodBody CreateThrowLinkedAwayBody (MethodDefinition method)
+		{
+			var body = new MethodBody (method);
+			var il = body.GetILProcessor ();
+
+			// import the method into the current assembly
+			var ctor = Context.MarkedKnownMembers.NotSupportedExceptionCtorString;
+			ctor = assembly.MainModule.ImportReference (ctor);
+
+			il.Emit (OpCodes.Ldstr, "Linked away");
+			il.Emit (OpCodes.Newobj, ctor);
+			il.Emit (OpCodes.Throw);
+			return body;
+		}
+
+		MethodBody CreateStubBody (MethodDefinition method)
+		{
+			var body = new MethodBody (method);
+
+			if (method.HasParameters && method.Parameters.Any (l => l.IsOut))
+				throw new NotImplementedException ();
+
+			var il = body.GetILProcessor ();
+			if (method.IsConstructor && !method.IsStatic) {
+				var base_ctor = GetDefaultInstanceConstuctor (method.DeclaringType.BaseType);
+				base_ctor = assembly.MainModule.ImportReference (base_ctor);
+
+				il.Emit (OpCodes.Ldarg_0);
+				il.Emit (OpCodes.Call, base_ctor);
+			}
+
+			switch (method.ReturnType.MetadataType) {
+			case MetadataType.Void:
+				break;
+			case MetadataType.Boolean:
+				il.Emit (OpCodes.Ldc_I4_0);
+				break;
+			default:
+				throw new NotImplementedException (method.ReturnType.FullName);
+			}
+
+			il.Emit (OpCodes.Ret);
+			return body;
+		}
+
+		static MethodReference GetDefaultInstanceConstuctor (TypeReference type)
+		{
+			foreach (var m in type.GetMethods ()) {
+				if (m.HasParameters)
+					continue;
+
+				var definition = m.Resolve ();
+				if (definition.IsStatic || !definition.IsConstructor)
+					continue;
+
+				return m;
+			}
+
+			throw new NotImplementedException ();
+		}
+
+		static void ClearDebugInformation (MethodDefinition method)
+		{
+			// TODO:
```

Comment:
I know it's possible it's just not what we should do (and you are not clean it up completely). I am waiting for better Cecil fix later

Reply:
I don't remember all the details, but I seem to recall that if we didn't at least clean up this much of the debug info that it would cause problems for something downstream in our workflow.  Maybe it was il2cpp.

Whatever the reason, I think adding that would be better than nothing.  You can leave the `//TODO` there as a reminder to clean it up once the new cecil api is available.

Reply:
Addressed